Don't query Sabre parent nodes

PVince81 opened the pull request:
## Description
When dealing with file access, the old DAV had a shortcut way to
directly access a file node without having to traverse all parents.
The new DAV endpoint needs this as well:

- added \OCA\DAV\Tree implementation with shortcut for "files"
- disable DAVACL plugin as it would also query all parents

There are several reasons for avoiding to query the parents:
- it causes file cache access for each node, not removable as it needs
to check for existence
- with external storages and update detection, querying a folder node
causes update detection to kick in and the scanner to do its work, the
scanner will set additional locks that might needlessly prevent
concurrent access.

The sad part is that any future Sabre plugin without caring for the above has a risk of retriggering these issues. So any Sabre plugin dealing with files must be aware of this and avoid needlessly querying every parent node.

## Related Issue
Fixes https://github.com/owncloud/core/issues/28779

## Motivation and Context
Read from https://github.com/owncloud/core/issues/28779#issuecomment-325927979.
But also this will improve performance or at least bring it closer to the one from the old DAV endpoint by avoiding additional file cache queries.

## How Has This Been Tested?
Test steps from above ticket: deletion of three folders simultaneously on an external storage with update detection kept enabled.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
Don't query Sabre parent nodes

When dealing with file access, the old DAV had a shortcut way to
directly access a file node without having to traverse all parents.
The new DAV endpoint needs this as well:

- added \OCA\DAV\Tree implementation with shortcut for "files"
- disable DAVACL plugin as it would also query all parents

There are several reasons for avoiding to query the parents:
- it causes file cache access for each node, not removable as it needs
to check for existence
- with external storages and update detection, querying a folder node
causes update detection to kick in and the scanner to do its work, the
scanner will set additional locks that might needlessly prevent
concurrent access.

jvillafanez commented:
> with external storages and update detection, querying a folder node
causes update detection to kick in and the scanner to do its work, the
scanner will set additional locks that might needlessly prevent
concurrent access.

This should be fine unless you're deleting a parent (at any level) of another folder that is being deleted at the same. In this case it's fine because there are operations that are happening in the contents of the folder.

I think just disabling the ACL plugin should be enough to solve the problem.

PVince81 commented:
> I think just disabling the ACL plugin should be enough to solve the problem.

It's not. I already tried: both fixes are needed else the 423 Locked exceptions will creep in.
This is because the object Tree also touches parent nodes

jvillafanez commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,138 @@
+		}
+
+		// check the path, also called when the path has been entered manually eg via a file explorer
+		if (\OC\Files\Filesystem::isForbiddenFileOrDir($path)) {
```
any reason for this duplication? it will a comment if any.

jvillafanez commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,138 @@
+		}
+
+		// Is it the root node?
+		if ($path === '') {
```
move this check up, before any complex operation?

jvillafanez commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,138 @@
+			throw new \Sabre\DAV\Exception\Forbidden();
+		}
+
+		$path = trim($path, '/');
```
we could move this before the `isForbiddenFileOrDir` check

PVince81 replied:
hah... seems it's also duplicated in ObjectTree...

jvillafanez commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,138 @@
+
+		$path = trim($path, '/');
+
+		if (isset($this->cache[$path]) && $this->cache[$path] !== false) {
```
this also might be possible to move it before the `isForbiddenFileOrDir` check, to return the cache data as soon as we can

PVince81 replied:
good idea

jvillafanez commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,138 @@
+
+		if (!$info) {
+			$this->cache[$path] = false;
+			throw new \Sabre\DAV\Exception\NotFound('File with name ' . $path . ' could not be located');
```
Do we want to cache this exception? I mean, if the file doesn't exists at some point, it won't likely exists during the rest of the request, so we might not want to go through all the steps again to get another NotFound. On the other hand, it might be fine since it's expected the exception isn't catched until the top level to send the message to the client.

PVince81 replied:
the exception is cached already with the value `false` for future checks

PVince81 commented:
@jvillafanez I actually copied getNodeForPath from the old ObjectTree::getNodeForPath and removed every bit related to chunking or part files. Then I added a few lines at the top to compute the correct path. The result is what you see above. 

I'll still have a look at your proposals for optimization, might be a good opportunity.

Before I do, do you agree with the general approach ?

I'm not sure if @DeepDiver1975 would appreciate that we brought back the Tree implementation but so far I don't see any better way. Too bad Sabre doesn't have a "getNodeForPath" event that plugins could override.

jvillafanez commented:
The only option left that I see in addition to this one is to override the RootCollection instead of the \OCA\DAV\Tree. I don't know what is the best option.

The other option would be to modify the View and change or add a method to get the file info without triggering a scan, but in addition to be risky (too many components might be affected) we'll need to evaluate what could be the consecuences of fetching not up-to-date data.

PVince81 commented:
> The only option left that I see in addition to this one is to override the RootCollection instead of the \OCA\DAV\Tree. I don't know what is the best option.

We already have our own RootCollection. Sadly the API `getChild()` only allows to get the next child, not a sub-sub-sub-child. And the calling code also calls it this way, recursively, step by step.

> The other option would be to modify the View and change or add a method to get the file info without triggering a scan, but in addition to be risky (too many components might be affected) we'll need to evaluate what could be the consecuences of fetching not up-to-date data.

I looked into this here: https://github.com/owncloud/core/pull/28855. You're right that it's risky so I'm less happy with https://github.com/owncloud/core/pull/28855 than with https://github.com/owncloud/core/pull/28853

PVince81 commented:
- [ ] adjust getNodeForPath based on comments
- [ ] fix unit tests
- [ ] write new unit tests for the new `Tree` class

PVince81 commented:
- [ ] double check that cached `false` will rethrow exception if queried again

PVince81 commented:
In theory we could invent a new interface that allows our sabre collections to provide "getNodeForPath" for a subpath. If implemented, call it. Else fall back to traversing with `getChild()`

PVince81 commented:
- [ ] double check `$this->tree->cacheNode()` usage in `Directory`, might not be using the correct path...

PVince81 commented:
We might not even need `cacheNode()` any more thanks to: https://github.com/fruux/sabre-dav/blob/master/lib/DAV/Tree.php#L200.

But I also found this... https://github.com/owncloud/core/pull/21401

PVince81 commented:
Now from reading the comments, it seems that doing the `cacheNode` thing was relevant mostly for the old DAV endpoint where we had the custom `ObjectTree`. Now in the new DAV endpoint there is no such tree. Well, there will be one added again from this PR, so not sure...

I think re-adding `getNodeForPath` is likely the correct way after reading this comment: https://github.com/owncloud/core/pull/21401#r49167496

PVince81 commented:
I have the feeling that this is going a bit too far already with risk of regressions or at least performance regressions... If I don't find a solution until next week I'll revert the web UI to use the old DAV endpoint for the time being.

SergioBertolinSG commented:
After mounting an external storage the web ui has an spinner spinning forever, refresh doesn't help.

SergioBertolinSG commented:
Retested, external storage is not required, it fails just opening files view:
```
{"reqId":"ZgobgAAahizRXXgw15PQ","level":3,"time":"2017-09-04T08:45:32+00:00","remoteAddr":"::1","user":"--","app":"PHP","method":"GET","url":"\/index.php","message":"Undefined variable: user at \/opt\/owncloud\/lib\/public\/Util.php#193"}
{"reqId":"yX6y7isiGhJ5tfbBkRDp","level":3,"time":"2017-09-04T08:46:16+00:00","remoteAddr":"HIDDEN_IP","user":"admin","app":"remote","method":"PROPFIND","url":"\/remote.php\/dav\/files\/admin\/","message":"Exception: {\"Exception\":\"TypeError\",\"Message\":\"Argument 2 passed to Sabre\\\\DAV\\\\Server::getPropertiesByNode() must implement interface Sabre\\\\DAV\\\\INode, null given, called in \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php on line 980\",\"Code\":0,\"Trace\":\"#0 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(980): Sabre\\\\DAV\\\\Server->getPropertiesByNode(Object(Sabre\\\\DAV\\\\PropFind), NULL)\\n#1 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1665): Sabre\\\\DAV\\\\Server->getPropertiesIteratorForPath('files\\\/admin', Array, 1)\\n#2 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(355): Sabre\\\\DAV\\\\Server->generateMultiStatus(Object(Generator), false)\\n#3 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpPropFind(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#5 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:PROPFIND', Array)\\n#6 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#7 \\\/opt\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(243): Sabre\\\\DAV\\\\Server->exec()\\n#8 \\\/opt\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#9 \\\/opt\\\/owncloud\\\/remote.php(175): require_once('\\\/opt\\\/owncloud\\\/a...')\\n#10 {main}\",\"File\":\"\\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php\",\"Line\":1056}"}
```

PVince81 commented:
argh... I was doing something experimental and moved around code but made the mistake to commit the removal of said code

PVince81 commented:
- [ ] cache the full DAV path in `Tree` instead of just the files subpath...

PVince81 commented:
@SergioBertolinSG please try again, thanks

PVince81 commented:
@jvillafanez mind rechecking ?

In general I'm not too happy about this but it feels like it's the best compromise so far.

I hope we can eventually get rid of `cachePath` which is called by `Directory::getChildren()' for every child. I believe that Sabre is supposed to have a way already to obsolete this, but the discussion from the ticket in https://github.com/owncloud/core/pull/21401#r49167496 makes me doubt about it and I don't want to change too much at once to avoid regressions. We really need to clean up this DAV mess at some point...

SergioBertolinSG commented:
Better than before.

While deleting several files at once there are locked files.

```
{"reqId":"N74dSbcSJPwWbc56Sjby","level":4,"time":"2017-09-04T10:36:13+00:00","remoteAddr":"HIDDEN_IP","user":"admin","app":"webdav","method":"DELETE","url":"\/remote.php\/dav\/files\/admin\/SFTP\/VARIAS_CARPETAS\/vikings-episode-4-trial%20copy%202.jpg","message":"Exception: {\"Message\":\"HTTP\\\/1.1 423 \\\"SFTP\\\/VARIAS_CARPETAS\\\/vikings-episode-4-trial copy 2.jpg\\\" is locked\",\"Exception\":\"OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Exception\\\\FileLocked\",\"Code\":0,\"Trace\":\"#0 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->delete()\\n#1 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('files\\\/admin\\\/SFT...')\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#5 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/opt\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(243): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/opt\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#8 \\\/opt\\\/owncloud\\\/remote.php(175): require_once('\\\/opt\\\/owncloud\\\/a...')\\n#9 {main}\",\"File\":\"\\\/opt\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/File.php\",\"Line\":363,\"User\":\"admin\"}"}
```

jvillafanez commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,157 @@
+		// use the cached version when needed
+		$filesRoot = parent::getNodeForPath('files/' . $userId);
+
+		$fileView = new View('/' . $userId . '/files/');
```
The view creation might not be very expensive, but I think it's better to create it just before being used. There are several cases where we don't need to create the view object.

jvillafanez commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,157 @@
+		}
+
+		if (!$info) {
+			$this->fileNodeCache[$cachePath] = false;
```
I'd rather have a clear expectation about missing fileInfos:
* If the expectation is NOT to cache the failure and let it retry afterwards, this isn't needed. In addition, the `$cachePath` variable isn't defined.
* If the expectation is to cache the failure so next call will return the same error, we should return the cached error as soon as possible.

If the valid results are being cached, I expect the errors are cached too. If this won't be the case, the PHPDoc should state that.

PVince81 replied:
yes, now throwing an exception if we detect a cached `false`

PVince81 commented on apps/dav/lib/Connector/Sabre/ObjectTree.php:
```diff
@@ -131,17 +131,17 @@ public function getNodeForPath($path) {
+			throw new \Sabre\DAV\Exception\Forbidden();
+		}
+
 		if ($path) {
```
this is copy-pasted from the old code... not intended. not calling verifyPath on "0" doesn't hurt though.

I'll fix this one here

PVince81 replied:
this check is useless as we now check for empty path above 🔫

PVince81 replied:
argh... I mixed up ObjectTree with Tree... I actually fixed this in Tree only

PVince81 commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,157 @@
+	/**
+	 * Cache a files specific node
+	 */
+	public function cacheNode(Node $node) {
```
It's an awful mess from the past: Directory::getChildren() uses it to cache children as it finds them.

It is likely not the right approach any more but I don't want to touch more code than necessary right now.

PVince81 replied:
https://github.com/owncloud/core/blob/master/apps/dav/lib/Connector/Sabre/Directory.php#L256

jvillafanez replied:
> It's an awful mess from the past: Directory::getChildren() uses it to cache children as it finds them.

:man_facepalming: Explain it in a comment for the time being, otherwise this will likely be deleted without mercy the next time, and something will explode.

Anyway, that's bad and we have to fix it at some point.

PVince81 commented:
@jvillafanez I fixed them all.

Now regarding @SergioBertolinSG's comment about the extra stray lock, I'll need to take some time to log all locks again and diff the lock traces between old dav and new dav. (this is how I found out how to fix this so far)

PVince81 commented:
I diff the lock order between old dav and new dav on this PR and they are identical:
```
acquireLock: type=1 path="a" 
acquireLock: type=1 path="" 
acquireLock: type=2 path="scanner::a" 
acquireLock: type=1 path="a" 
acquireLock: type=1 path="a" 
releaseLock: type=1 path="a" 
acquireLock: type=1 path="a\/data1.dat" 
releaseLock: type=1 path="a\/data1.dat" 
releaseLock: type=1 path="a" 
releaseLock: type=2 path="scanner::a" 
releaseLock: type=1 path="a" 
releaseLock: type=1 path="" 
acquireLock: type=1 path="a" 
acquireLock: type=1 path="" 
changeLock: type=2 path="a" 
changeLock: type=1 path="a" 
acquireLock: type=1 path="a" 
acquireLock: type=1 path="" 
releaseLock: type=1 path="a" 
releaseLock: type=1 path="" 
releaseLock: type=1 path="a" 
releaseLock: type=1 path="" 
acquireLock: type=1 path="" 
acquireLock: type=2 path="scanner::" 
acquireLock: type=1 path="" 
acquireLock: type=1 path="" 
releaseLock: type=1 path="" 
acquireLock: type=1 path="x" 
releaseLock: type=1 path="x" 
acquireLock: type=1 path="b" 
releaseLock: type=1 path="b" 
acquireLock: type=1 path="c" 
releaseLock: type=1 path="c" 
releaseLock: type=1 path="" 
releaseLock: type=2 path="scanner::" 
releaseLock: type=1 path="" 
```

so if there is a stray lock somewhere it was probably already there with the old dav endpoint, so probably a different issue.

PVince81 commented:
I have uploaded the scan log results to https://cloud.owncloud.com/index.php/s/VK1ncpEUdvhScz9 as "log-delete-fixed*". If you diff the logs and ignore the stack traces, they are identical.

jvillafanez commented:
I don't see the changes 🙈

PVince81 commented:
@jvillafanez I did push, here is one change for example: https://github.com/owncloud/core/pull/28853/files#diff-d4fb9df74a6b54801784acb672a1a952R112

What changes are you missing ?

I'll process your new comments

PVince81 commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,157 @@
+	 * @param \Sabre\DAV\INode $rootNode
+	 */
+	public function __construct(ICollection $rootNode) {
+		$this->rootNode = $rootNode;
```
will do

PVince81 commented:
@jvillafanez adjusted again and squashed

PVince81 commented:
```
10:12:00 1) OCA\DAV\Tests\unit\ServerTest::test
10:12:00 ArgumentCountError: Too few arguments to function Sabre\DAV\Tree::__construct(), 0 passed in /var/lib/jenkins/workspace/owncloud-core_core_PR-28853-LZGDUAKIZIJB37QSAQ64PKSB6ATNPVHHK6ENRW2HVRSYU5ZK3ORA/apps/dav/lib/Tree.php on line 61 and exactly 1 expected
10:12:00 
10:12:00 /var/lib/jenkins/workspace/owncloud-core_core_PR-28853-LZGDUAKIZIJB37QSAQ64PKSB6ATNPVHHK6ENRW2HVRSYU5ZK3ORA/lib/composer/sabre/dav/lib/DAV/Tree.php:41
10:12:00 /var/lib/jenkins/workspace/owncloud-core_core_PR-28853-LZGDUAKIZIJB37QSAQ64PKSB6ATNPVHHK6ENRW2HVRSYU5ZK3ORA/apps/dav/lib/Tree.php:61
10:12:00 /var/lib/jenkins/workspace/owncloud-core_core_PR-28853-LZGDUAKIZIJB37QSAQ64PKSB6ATNPVHHK6ENRW2HVRSYU5ZK3ORA/apps/dav/lib/Server.php:70
10:12:00 /var/lib/jenkins/workspace/owncloud-core_core_PR-28853-LZGDUAKIZIJB37QSAQ64PKSB6ATNPVHHK6ENRW2HVRSYU5ZK3ORA/apps/dav/tests/unit/ServerTest.php:42
```

PVince81 commented:
☕️ had no effect

Fixed issue, rechecked web UI and pushed/squashed again.

jvillafanez commented:
Waiting for tests results....

PVince81 commented:
I had a quick test with the calendar app to see if the caldav endpoint still works, and also with the comments app to verify that non-files endpoints still work.

PVince81 commented:
okay, we got some stuff...

``` 
    checksums.Uploading a file with checksum should work using new dav path
    checksums.Uploading a file with checksum should return the checksum in the propfind using new dav path
    checksums.Uploading a file with checksum should return the checksum in the download header using new dav path
    checksums.Moving a file with checksum should return the checksum in the propfind using new dav path
    checksums.Moving file with checksum should return the checksum in the download header using new dav path
    checksums.Copying a file with checksum should return the checksum in the propfind using new dav path
    checksums.Copying file with checksum should return the checksum in the download header using new dav path
    checksums.Upload new dav chunked file where checksum matches
    quota.Uploading a file as owner having enough quota
    quota.Overwriting a file as owner having enough quota
    quota.Uploading a file in received folder having enough quota
    quota.Overwriting a file in received folder having enough quota
    quota.Overwriting a received file having enough quota
    trashbin-new-endpoint.deleting a file of a shared folder moves it to trashbin
    trashbin-new-endpoint.deleting a shared folder moves it to trashbin
    trashbin-new-endpoint.deleting a received folder doesn't move it to trashbin
    trashbin-new-endpoint.deleting a file in a received folder it is moved to the trashbin
    trashbin-new-endpoint.deleting a file in a received folder when restored it comes back to the original path
    trashbin-new-endpoint.Deleting a folder into external storage moves it to the trashbin
    trashbin-new-endpoint.Deleting a file into external storage moves it to the trashbin and can be restored
    webdav-related-new-endpoint.Moving a file
    webdav-related-new-endpoint.Setting custom DAV property and reading it after the file is renamed
    webdav-related-new-endpoint.Upload file via new chunking endpoint with correct size header
```

PVince81 commented:
@jvillafanez oh no... can of worms mode engaged...

Remember that we said that if a file's non-existence is cached as `false`, we should rethrow the exception when queried again ? It makes sense from an API point of view. Unfortunately it is this change that breaks the integration tests.

I'll have a look what the real Sabre's ObjectTree is supposed to do.

PVince81 commented:
this is very fishy... I see that the regular Sabre Tree which was used before this PR does throw an exception here: https://github.com/fruux/sabre-dav/blob/3.2.2/lib/DAV/Tree.php#L74.
It doesn't catch the error, so I expect that subsequent calls would still throw the exception.

And if it doesn't reach this bit, then `getChild` is the one that would throw the exception, which we do correctly here: https://github.com/owncloud/core/blob/master/apps/dav/lib/Connector/Sabre/Directory.php#L247

(a side note: I see that `Directory::getChild()` is missing a call to `$this->tree->cacheNode(false)` in case of non-existence)

PVince81 commented:
raised https://github.com/owncloud/core/issues/28912 to gather the horrors we find

jvillafanez commented:
> Remember that we said that if a file's non-existence is cached as false, we should rethrow the exception when queried again ? It makes sense from an API point of view. Unfortunately it is this change that breaks the integration tests.

Not caching the exception migth be fine: you query the file and it isn't present, then query again and now it is present because the file is created between those calls.
This is fine as long as we clearly specify that this is the expected behaviour.

PVince81 commented:
from the checksum test, this is the stack that would trigger the second throw from a cached `false`:
```
0  OCA\DAV\Tree->getNodeForPath() /srv/www/htdocs/owncloud/apps/dav/lib/Tree.php:118
1  OCA\DAV\Connector\Sabre\FilesPlugin->sendFileIdHeader() /srv/www/htdocs/owncloud/apps/dav/lib/Connector/Sabre/FilesPlugin.php:404
2  call_user_func_array:{/srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
3  OCA\DAV\Connector\Sabre\Server->emit() /srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
4  OCA\DAV\Connector\Sabre\Server->createFile() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/Server.php:1101
5  Sabre\DAV\CorePlugin->httpPut() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/CorePlugin.php:525
6  call_user_func_array:{/srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
7  OCA\DAV\Connector\Sabre\Server->emit() /srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
8  OCA\DAV\Connector\Sabre\Server->invokeMethod() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/Server.php:479
9  OCA\DAV\Connector\Sabre\Server->exec() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/Server.php:254
10 OCA\DAV\Server->exec() /srv/www/htdocs/owncloud/apps/dav/lib/Server.php:243
11 require_once()  /srv/www/htdocs/owncloud/apps/dav/appinfo/v2/remote.php:31
12 {main}          /srv/www/htdocs/owncloud/remote.php:175
```

Apparently the file was created and exists now, but our current cache still contains the old value `false`. The old code path would go through the View and calls `file_exists` there, so it works.

The default Sabre `Tree:nodeExists()` doesn't go to the cache and also goes to the first parent with `getNodeForPath()` then calls `childExists()`. This is why we don't bump into that issue.

This all still looks fishy, even in the Sabre code. I could imagine scenarios where a cached node used to exist but was deleted. And I don't see any unbind events registered there. :sad:

PVince81 commented:
listening to bind/unbind events might be a solution, but I see that code like `httpPut` doesn't even trigger a `beforeBind` event.

Ok, at least a deletion calls `Tree:delete()` from Sabre which calls `markDirty` which invalidates the cache. Then I'd need to override it to also invalidate the `fileNodeCache`.

- [ ] implement markDirty to invalidate `fileNodeCache` as we need to map the absolute path to the relative path

Maybe we should make the `FilesHome` actually extend `Sabre\DAV\Tree` and also implement the `ICollection` methods additionally. Then make our base Tree just check if the parent node we got `instanceof Tree` and then call `getNodeForPath` on the remaining subtree.

I have a bad feeling about this... The more we touch the more will break. I'm tending more in the direction of deferring this to 10.0.4.

PVince81 commented:
If we had separate Sabre servers for each of the subtrees "/files", "/comments", etc, this wouldn't be an issue. It would even have the advantage that each subtree would register only the required plugins instead of everything.

PVince81 commented:
Raised https://github.com/owncloud/core/issues/28913 to reconsider and discuss separate trees/servers for each first-level node.

PVince81 commented:
For now I'm very unhappy about having to map paths every time in the base `Tree`. It's already ugly that the base `Tree` needs to know anything about the files endpoint and potentially others in the future.

So, I'm going to explore the following:
- [ ] make FilesHome extend `\Sabre\DAV\Tree` and implement `ICollection` (basically bringing back the old ObjectTree but on a sub-level)
- [ ] root `Tree` implementation where `getNodeForPath()` goes down from the root with `getChild()` iterations, for each iteration checking if the node is a `Sabre\DAV\Tree`

PVince81 commented:
or I just make the current `Tree` work... seems our old tree overwrote `nodeExists` with this: https://github.com/owncloud/core/blob/v10.0.3beta2/apps/dav/lib/Connector/Sabre/ObjectTree.php#L110

Commit:
DAV FilesHome is now also a tree

PVince81 commented:
@jvillafanez check this out: https://github.com/owncloud/core/pull/28853/commits/737c9b03033cf06360ac80c966df090500699cc3

At least now the specialized files code is in `FilesHome` which extends the old `ObjectTree`.
Whether we agree to reuse the old `ObjectTree` here in new dav is something to discuss...
This seems to solve the checksum errors, we'll see if other tests also pass now.

Now what's open for later is that the old `ObjectTree` implementation isn't consistent in regard to handling non-existing nodes and throwing exceptions, something to look into later... At least now we only have one occurence of the "bad code" within `ObjectTree` instead of having another copy in `Tree`.

@jvillafanez let me know what you think.

PVince81 commented:
- [ ] adjust TreeTest to test the new logic

jvillafanez commented:
I'm not sure if the approach is good or not. Is there any conceptual difference between a directory and an object tree? What is expected to be attached as FilesHome? I mean I expect DAV to manage nodes, so if I query the root collection to get the expected FilesHome, should I get a node or a tree?

In addition, the constructor of the FilesHome feels quite forced: if we need a root node, why don't we inject it in the constructor instead of creating it there? Maybe we need to create a new class for that and use this new class for the new webdav endpoint.

PVince81 commented:
@jvillafanez I think Sabre's concept is that basically there is only one tree which sits at the root and embeds the root node. Then each node is just an implementation of `INode`.

Here we're innovating a bit because we're implementing another tree in the middle of the main tree.

Injecting a root node in the constructor feels wrong, because it would have to be injected by the parent which is the node for "/files/$userId". Well, even if we do it there, we'd anyway need to instantiate a `Directory` somewhere in the tree. We can't inject it from the root. I think it is ok for `FilesHome` to contain/wrap the node like it does here.

I want to discuss this further with @DeepDiver1975 later to make sure we're all on the same page regarding this "innovative" usage.

PVince81 commented:
now that we switched back to old DAV, we'll need this PR https://github.com/owncloud/core/pull/28874 to reproduce the original issue this PR solves

PVince81 commented:
- [ ] BUG:
```
{"reqId":"KSPBceR23bRLjz65nvor","level":3,"time":"2017-10-12T14:37:31+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"PHP","method":"PROPFIND","url":"\/owncloud\/remote.php\/dav\/files\/admin\/Photos\/","message":"Declaration of OCA\\DAV\\Files\\FilesHome::getChildren() should be compatible with Sabre\\DAV\\Tree::getChildren($path) at \/srv\/www\/htdocs\/owncloud\/apps\/dav\/lib\/Files\/FilesHome.php#30"}
```

DeepDiver1975 commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,87 @@
+	 * @throws \Sabre\DAV\Exception\Forbidden
+	 * @throws \Sabre\DAV\Exception\ServiceUnavailable
+	 */
+	public function getNodeForPath($path) {
```
Isn't the parent implementation performing a proper caching?
why not?
shall we fix this in the parent class?

PVince81 replied:
The parent impl is caching yet. The problem is that we should not query the parents **at all** as they are not useful for our current operation. And even with caching, the first time query is still being done for the parents which triggers our internal View APIs which triggers the scanning.

PVince81 replied:
Basically the approach we had in the old endpiont: getNodeForPath goes **directly** to the target node without traversing any parents, as we know that all parents are filesystem folders.

With the default Sabre approach as used by the new endpoint, getNodeForPath first traverses all parents recursively up to the root and caches these, and then calls `getChild()` on every node until reaching the actually queried node.

The returned result is the same, except that in the latter case we traversed Folder nodes and had to query getFileInfo() on every parent while in the former we only call getFileInfo() on the final child node.

Hope this makes sense.

PVince81 commented:
@mrow4a might be interested in this ? At least come here if you happen to observe the perf regression on the new DAV endpoint

PVince81 commented:
moving to planned for 10.0.5.

I hope I can get your feedback and understanding by then @DeepDiver1975

DeepDiver1975 commented on apps/dav/lib/Files/FilesHome.php:
```diff
@@ -38,10 +40,37 @@ class FilesHome extends Directory {
+	}
+
+	function createDirectory($name) {
+		return $this->rootNode->createDirectory($name);
```
remove return - createDirectory returns void

DeepDiver1975 commented on apps/dav/lib/Files/FilesHome.php:
```diff
@@ -38,10 +40,37 @@ class FilesHome extends Directory {
+		return $this->rootNode->getChild($name);
+	}
+
+	function getChildren() {
```
Tree::getChildren($path) vs ICollection::getChildren - this is not clean

DeepDiver1975 commented on apps/dav/lib/Files/FilesHome.php:
```diff
@@ -38,10 +40,37 @@ class FilesHome extends Directory {
+		return $this->rootNode->getLastModified();
 	}
 
 	function delete() {
```
Tree::delete($path) vs INode::delete

DeepDiver1975 commented on apps/dav/lib/Tree.php:
```diff
@@ -0,0 +1,87 @@
+
+namespace OCA\DAV;
+
+use OCA\DAV\Connector\Sabre\Exception\Forbidden;
```
many unneeded use statements

DeepDiver1975 commented on apps/dav/tests/unit/TreeTest.php:
```diff
@@ -0,0 +1,174 @@
+namespace OCA\DAV\Tests\unit;
+
+
+use OC\Files\FileInfo;
```
unneeded use statements

DeepDiver1975 commented on apps/dav/tests/unit/TreeTest.php:
```diff
@@ -0,0 +1,174 @@
+		$outputFileName,
+		$type
+	) {
+		$rootNode = $this->getMockBuilder('\OCA\DAV\Connector\Sabre\Directory')
```
Directory::class

PVince81 replied:
ok, so I copy-pasted this from the old code. I can clean this up.

DeepDiver1975 commented on apps/dav/tests/unit/TreeTest.php:
```diff
@@ -0,0 +1,174 @@
+			->with($fileInfoQueryPath)
+			->will($this->returnValue($fileInfo));
+
+		$tree = new \OCA\DAV\Connector\Sabre\ObjectTree();
```
use ....

DeepDiver1975 commented on apps/dav/tests/unit/TreeTest.php:
```diff
@@ -0,0 +1,174 @@
+
+		$storage = new Temporary([]);
+
+		$view = $this->createMock('\OC\Files\View', ['resolvePath']);
```
createMock only takes one arg

DeepDiver1975 commented on apps/dav/tests/unit/TreeTest.php:
```diff
@@ -0,0 +1,174 @@
+		$mountManager = $this->createMock('\OC\Files\Mount\Manager');
+
+		$tree = new \OCA\DAV\Connector\Sabre\ObjectTree();
+		$tree->init($rootNode, $view, $mountManager);
```
please add phpdoc to $rootNode, $view, $mountManager to define theirs type

```php
/** @var Directory | \PHPUnit_Framework_MockObject_MockObject $rootNode */

DeepDiver1975 commented:
I'm okay with this change - please take care about my comments - THX

PVince81 commented:
Thanks. Glad that we agreed on the approach.

I'll take care of the comments next and unit tests and codecov.

DeepDiver1975 committed:
Add more tests

PVince81 commented:
Transferred to @DeepDiver1975 for hijacking

PVince81 commented:
:+1: for @DeepDiver1975's changes

PVince81 commented:
I see some tab issues, fixing... (copy-pasted code from Sabre that has spaces instead of tabs)

PVince81 commented:
I've squashed the tab fixes into the second commit and rebased onto master.

If CI passes let's merge this and backport.
